File: bill_splitter_python_algorithm/bill_splitter.py
def bill_splitter(list):
    total = sum(list)
    items = len(list)
    avg = total / items
    diff_list = [x - avg for x in sorted(list)]
    pos_count = sum(1 for x in diff_list if x >= 0)
    neg_count = len(diff_list) - pos_count

    transactions = []
    i, j = 0, len(diff_list) - 1

    while i < j:
        while diff_list[i] < 0:
            if -diff_list[i] > diff_list[j]:
                amount = round(diff_list[j], 2)
                diff_list[i] += diff_list[j]
                diff_list[j] = 0
                if amount > 0:
                    transactions.append((i, j, amount))
                j -= 1
            else:
                amount = round(-diff_list[i], 2)
                diff_list[j] += diff_list[i]
                diff_list[i] = 0
                if amount > 0:
                    transactions.append((i, j, amount))
        i += 1

    return transactions

lst = [71, 159, 142, 100, 69, 100, 12, 2]

print(bill_splitter(lst))

File: src/billSplitter.js
const splitBill = (data) => {
  const newData = data
    .map((item) => ({
      ...item,
      amount: parseFloat(item.amount),
    }))
    .sort((a, b) => a.amount - b.amount);

  const total = newData.reduce(
    (total, person) => total + person.amount || 0,
    0
  );

  const items = newData.length;
  const avg = total / items;

  const diffList = newData.map((item) => ({
    ...item,
    amount: item.amount - avg,
  }));

  let transactions = [];
  let i = 0;
  let j = diffList.length - 1;
  let sender;
  let recipient;
  let amount;

  while (i < j) {
    sender = diffList[i];
    while (sender.amount < 0) {
      recipient = diffList[j];
      if (-sender.amount > recipient.amount) {
        amount = parseFloat(recipient.amount.toFixed(2));
        sender.amount += recipient.amount;
        recipient.amount = 0;
        if (amount > 0) {
          transactions.push({
            sender: sender.name,
            recipient: recipient.name,
            amount: amount,
          });
        }
        j -= 1;
      } else {
        amount = -sender.amount.toFixed(2);
        recipient.amount += sender.amount;
        sender.amount = 0;
        if (amount > 0) {
          transactions.push({
            sender: sender.name,
            recipient: recipient.name,
            amount: amount,
          });
        }
      }
    }
    i += 1;
  }

  return transactions;
};

export { splitBill };

// Data sample in case sth went wrong
// const data = [
//   { name: 'John', amount: '12.45' },
//   { name: 'Bob', amount: '13.56' },
//   { name: 'Steven', amount: '1.43' },
//   { name: 'Kate', amount: '7.98' },
// ];

// splitBill(data)
